### ISSUE TITLE
[Browser Bookmarks] ...

### ISSUE THREAD

**Issue by lovrobuday** [raycast/extensions#10584]
### Extension

https://www.raycast.com/raycast/browser-bookmarks

### Description

Love this thing, but the more bookmarks I added the worse the search is. I named my bookmarks like "work live" but when I search "work live" that bookmark is 5th on the list

### Who will benefit from this feature?

search based on bookmarks name rather than url, better search

**Comment by thomaslombart** [comment]
It's tricky to get a good fuzzy search experience in the browser's bookmarks. Which results do you get before the one you want with which query?

**Comment by lovrobuday** [comment]
So the query is:
covea live

The results:
jira - TimeSheet
cover staging
cover live
...

the first result does have a word that starts with "co" and a word that starts with "li" one after the other in the url which might be the issue, and cover staging has "cover" in the url where cover live doesnt, but it would be nice to have it fuzzy find over the bookmark name rather than url

**Comment by thomaslombart** [comment]
Do you open "Jira - TimeSheet" a lot? When searching, the extension will always show first the bookmarks you open more frequently than the others. Otherwise, we do have fuzzy search over the bookmark name with a greater weight than the URL (as you can see [here](https://github.com/raycast/extensions/blob/05c4ea93c2bd362a4c031321faca83b93f103b63/extensions/browser-bookmarks/src/index.tsx#L260)). Improving the fuzzy search functionality is challenging as modifying certain settings may affect other users.

**Comment by dariodjuric** [comment]
> Do you open "Jira - TimeSheet" a lot? When searching, the extension will always show first the bookmarks you open more frequently than the others.

For me, the bookmarks search was always appearing in the same order, even though I would use some bookmarks more than others. Then I removed Raycast using Appcleaner and installed it again, now the bookmarks get sorted based on the frequency.

**Comment by dariodjuric** [comment]
Perhaps a better solution would be to not consider _all_ results before sorting by frecency. In my case I have a total of 71 results for that search term, so if one bookmark with a really low score but a really high freqency gets into the result, it will appear at the top, which I don't expect.

Something like this, perhaps?


```ts
  const fuse = useMemo(() => {
    return new Fuse(folderBookmarks, {
      keys: [
        { name: "title", weight: 3 },
        { name: "domain", weight: 1 },
        { name: "folder", weight: 0.5 },
      ],
      threshold: 0.4,
      includeScore: true,
      ignoreLocation: true,
      shouldSort: true, // <-- sort the results
    });
  }, [folderBookmarks]);

  const filteredBookmarks = useMemo(() => {
    if (query === "") {
      return folderBookmarks;
    }

    const searchResults = fuse.search(query);

    return searchResults.slice(0, 10) // <-- limit the results before doing further sorting
```

**Comment by thomaslombart** [comment]
I get where you're coming from but I got opposite feedback in the past about that. Some people expect results with a high frecency to be sorted before others, even if the first result is less relevant to the query than the rest. So making changes here will definitely impact others.

It seems to me that the issue here is that the query matches the folder name even though it shouldn't. Maybe we can match exact words for folders instead of a fuzzy search?

**Comment by dariodjuric** [comment]
> I get where you're coming from but I got opposite feedback in the past about that. Some people expect results with a high frecency to be sorted before others, even if the first result is less relevant to the query than the rest. So making changes here will definitely impact others.

I see. Do you think it makes sense to make this configurable in the extension's UI so that everyone can tweak as needed?

E.g. have the following option:

- Prefer Better Match: true/false (Prefer better matching bookmarks before sorting by frecency.)

Or maybe even these for extra customization:

- Search by Domain: true/fasle
- Search by Folder Names: true/false

(I can make a PR if it helps.)

> It seems to me that the issue here is that the query matches the folder name even though it shouldn't. Maybe we can match exact words for folders instead of a fuzzy search?

Hm, I'm not sure if that will fix the OP's issue, but for me that would fix this particular case because less bookmarks would appear as relevant.

**Comment by thomaslombart** [comment]
The tricky bit is that it often depends on whatever you're searching for. Searching by domain may make sense for a certain search while others don't. And configuring the settings takes more time than the search itself.

I'd love to find an algorithm that's good enough for all use-cases because Fuse.js (what we use for Fuzzy search) doesn't really work out here. I'll discuss that with the team to see how we can improve it.

**Comment by dariodjuric** [comment]
Sounds good! Just to clarify, I was referring to the settings on the extension level (like where the "Show Bookmark Domain" setting is). As different people have different expectations on how the search would behave, they would be able to configure the extension based on those expectations.

Thanks for looking into this! Hoping we'll find a solution that fits everyone.

**Comment by dariodjuric** [comment]
Here's another example from today where a high-score result gets before the result I was searching for:

<img width="745" alt="CleanShot 2024-03-11 at 08 32 18@2x" src="https://github.com/raycast/extensions/assets/5459296/5630593a-c7a0-4950-8dbe-025b2a63ffef">

**Comment by ksandvik** [comment]
I would say turn off any fuzzy search and just search on strings and sub-strings, there's  extremely little benefit of fuzzy searching and not getting results for defined sub-strings is a far worse case.

**Comment by thomaslombart** [comment]
@ksandvik Thanks for your input! Feel free to submit a PR if you have any time for this 🙂

**Comment by niksite** [comment]
> I would say turn off any fuzzy search

Yeah! Making fuzzy search optional would resolve all those issues.
